instance/tests/management/test_instance_statistics_csv.py

```diff
@@ -0,0 +1,252 @@
+    """
+    @freezegun.freeze_time('2020-01-01')
+    def test_invalid_input(self):
+        """
+        Verify that invalid date strings error and provide reasonable message
+        """
+        with self.assertRaises(ArgumentTypeError) as context_manager:
+            valid_date('invalid')
+            self.assertIn('Not a valid date', str(context_manager.exception))
+
+    @freezegun.freeze_time('2020-01-01')
+    def test_date_in_future(self):
+        """
+        Verify that dates in the future error and provide reasonable message
+        """
+        with self.assertRaises(ArgumentTypeError) as context_manager:
+            valid_date('2020-05-01')
+            self.assertIn('Date must be earlier than or equal to today', str(context_manager.exception))
+
+    @freezegun.freeze_time('2020-01-01')
+    def test_today(self):
+        """
+        Verify that today works
+        """
+        self.assertEqual(valid_date('2020-01-01'), '2020-01-01')
+
+    @freezegun.freeze_time('2020-01-01')
+    def test_date_in_past(self):
+        """
+        Verify that a date in the past works
+        """
+        self.assertEqual(valid_date('2019-01-01'), '2019-01-01')
+
+
+class InstanceStatisticsCSVTestCase(TestCase):
+    """
+    Test cases for the `instance_statistics_csv` management command.
+    """
+    def test_no_qualified_domain(self):
+        """
+        Verify that the command correctly notifies the user that the domain is required
+        """
+        with self.assertRaises(CommandError) as context_manager:
+            call_command(
+                'instance_statistics_csv'
+            )
+            self.assertIn('Error: the following arguments are required', str(context_manager.exception))
+
+    @patch('instance.management.commands.instance_statistics_csv.open', Mock(side_effect=PermissionError))
+    def test_unable_to_write_to_output(self):
+        """
+        Verify that the command correctly notifies the user when it is unable to write to the request output file
+        """
+        stderr = StringIO()
+        with self.assertRaises(SystemExit):
+            call_command(
+                'instance_statistics_csv',
+                domain='test',
+                out='does_not_exist.txt',
+                stderr=stderr
+            )
+        self.assertIn('Permission denied while attempting to write file', stderr.getvalue())
+
+    def test_no_matching_instances(self):
+        """
+        Verify that the command correctly notifies the user that
+        there are no matching instances.
+        """
+        stderr = StringIO()
+        with self.assertRaises(SystemExit):
+            call_command(
+                'instance_statistics_csv',
+                domain='test.opencraft.hosting',
+                stderr=stderr
+            )
+        self.assertIn('No OpenEdXInstance exists with an external or internal domain', stderr.getvalue())
+
+    @patch('instance.models.openedx_instance.OpenEdXInstance.objects')
```

Comment:
@pcockwell, you can create an instance using the `OpenEdXInstanceFactory` and then not spawn an appserver before testing this scenario instead of mocking the model manager.

Reply:
Done in next commit